cql3: cartesian product limits

avikivity opened the pull request:
Cartesian products (generated by IN restrictions) can grow very large, even for short queries. This can overwhelm server resources.

Add limit checking for cartesian products, and configuration items for users that are not satisfied with the default of 100 records fetched.

Fixes #4752.

Tests: unit (dev), manual test with SIGHUP.

avikivity commented:
Note: do not merge, depends on seastar patch to add sharded<something_with_no_stop_method>

haaawk commented on db/config.cc:
```diff
@@ -674,6 +674,12 @@ db::config::config(std::shared_ptr<db::extensions> exts)
+    , max_partition_keys_per_query(this, "max_partition_keys_per_query", liveness::LiveUpdate, value_status::Used, 100,
+            "Maximum number of distinct partition keys per query. This limit places a bound on the size of IN tuples, "
+            "especially when multiple partition key columns have IN restrictions. Increasing this value can result in server instability.")
+    , max_clustering_keys_per_query(this, "max_clustering_keys_per_query", liveness::LiveUpdate, value_status::Used, 100,
```
This name is a bit misleading it may create an impression that it will limit the number of keys returned by a query with no restrictions. I know there's a description but maybe a better name would help avoid users misunderstand the option.

avikivity replied:
Yes, I thought of that too but did not find a better name.

avikivity replied:
@glommer perhaps you have a suggestion for a better name?

glommer replied:
why not `max_partition_restrictions_per_in_query` ?

avikivity replied:
I don't like it (nobody will guess what restrictions are, it's an internal term) but it's better that what we have no so I'll change to it.

haaawk commented on main.cc:
```diff
@@ -662,12 +663,14 @@ int main(int ac, char** av) {
             static sharded<db::system_distributed_keyspace> sys_dist_ks;
             static sharded<db::view::view_update_generator> view_update_generator;
+            static sharded<cql3::cql_config> cql_config;
+            cql_config.start().get();
```
Should we stop cql_config at some point?

avikivity replied:
Since we don't shut down everything, this carries the risk that system queries executed after cql_config is shut down will reference deallocated memory.

avikivity replied:
(clean shutdown is an all-or-nothing thing)

avikivity commented on cql3/query_options.hh:
```diff
@@ -55,6 +55,9 @@
 namespace cql3 {
 
+class cql_config;
+extern const cql_config default_cql_config;
```
It is used in query_options.cc and trace_keyspace_helper.cc.

avikivity commented on init.cc:
```diff
@@ -37,9 +37,11 @@ logging::logger startlog("init");
 
-void init_storage_service(distributed<database>& db, sharded<gms::gossiper>& gossiper, sharded<auth::service>& auth_service, sharded<db::system_distributed_keyspace>& sys_dist_ks,
+void init_storage_service(distributed<database>& db, sharded<gms::gossiper>& gossiper, sharded<auth::service>& auth_service,
+        sharded<cql3::cql_config>& cql_config,
```
I wanted to do the minimal change so it is easy to review. I can reformat to one parameter per line, perhaps in a follow-up, if it is necessary.

dekimir replied:
We shouldn't need these kinds of contortions with GitHub's patch viewer.  Even if you reformat in this PR, I bet it'll make it easy to read (like it did [here](https://github.com/dekimir/RamFuzz/commit/b8b9b101a16ca16ce62160b67ac2be64c0e0c49f#diff-9fa0c8bbb45efb32dfa3157ba20791f6L109), for example).

avikivity replied:
(suddenly off-topic comments are okay)

I can't always view patches in github's patch viewer. Much of my work is done from the command line.

dekimir replied:
Not off-topic; this concerns your inconsistent indentation, which you claim makes this (PR?) easy to review.  I'm just wondering if that's actually correct.

avikivity replied:
It's easier to review a small change than a larger change. Highlighting the differences makes the larger change easier to review, but a smaller change is easier still.

avikivity commented on tests/cql_query_test.cc:
```diff
@@ -365,6 +365,34 @@ SEASTAR_TEST_CASE(test_in_clause_validation) {
     });
 }
 
+SEASTAR_THREAD_TEST_CASE(test_in_clause_cartesian_product_limits) {
```
This does not generate a cartesian product, because partition keys and clustering keys are queried in separate steps.

denesb commented on cql3/restrictions/single_column_primary_key_restrictions.hh:
```diff
@@ -46,6 +46,8 @@
 #include "cartesian_product.hh"
 #include "cql3/restrictions/primary_key_restrictions.hh"
 #include "cql3/restrictions/single_column_restrictions.hh"
+#include "restrictions_config.hh"
```
`cql3/cql_config.hh` already includes this, why does it need to be included again here?

avikivity replied:
I can remove it.

denesb commented on cql3/restrictions/single_column_primary_key_restrictions.hh:
```diff
@@ -55,6 +57,14 @@ namespace cql3 {
+void check_cartesian_product_size(size_t size, size_t max) {
+    if (size > max) {
+        throw std::runtime_error(fmt::format("Primary key cartesian product size {} is greater than maximum {}",
+                size, max));
```
Wouldn't it be useful to know, which kind of key (partition or clustering) is this? Can we ever have both in a single query? (I think we can).

avikivity replied:
Yes

avikivity replied:
added

avikivity commented:
Updated, please re-review:

 - remove unnecessary include
 - add anonymous namespace for implementation names
 - report which restricted component limit we are exceeding
 - rebased around merged Seastar dependency
 - renamed config variables

avikivity commented:
@glommer I see you are marked as a reviewer

glommer commented on tests/cql_query_test.cc:
```diff
@@ -365,6 +365,34 @@ SEASTAR_TEST_CASE(test_in_clause_validation) {
+                        std::runtime_error);
+    }).get();
+}
+
```
One of the issues we had that led to the development of this patch was a single partition key being restricted.
Last we spoke I remember you told me that this would generate the same code paths in the end, despite not being a cartesian product.

In that case shouldn't we add a test case?

My fear is that someone may in the future change that and dissociate the code paths in which case queries that were invalid will magically become valid, despite having the same number of restrictions

avikivity replied:
Sure.

avikivity replied:
Added 4 cases: {pass, fail} x {one partition key, one clustering key}

glommer commented:
I had a small comment only

avikivity commented:
Updated with 4 more test cases as per @glommer request.

glommer commented:
@avikivity what is the current status of this ?

tgrabiec commented:
Needs to be rebased to resolve conflicts

avikivity commented:
rrrebased and rrretested

avikivity commented:
Merged as 080989d29624aeced6fab113ebdcbfa6e85be57b
